mne/io/edf/edf.py

```diff
@@ -208,7 +208,7 @@ def __init__(
                 tal_data[0],
                 encoding=encoding,
             )
-            self.set_annotations(annotations)
+            self.set_annotations(annotations, on_missing="warn")
```

Comment:
This is not gonna create annotations which do not coincide with the channel names as you wrote in the description, but it will completely ignore those annotations (and they will thus be absent from the `Raw` object entirely).

I'm not entirely sold on this fix, even if I proposed it. Especially after this sentence from @tomasdominik

| my EEG device records bipolar channels referenced to CPz, so the default name for the Fp1 channel, for example, is "EEG Fp1-CPz"

How come the recorded channel name is `"EEG Fp1-CPz"` and the annotation has a different channel name `"Fp1"`? Does that mean we erroneously parse the annotation read from an EDF file?

Reply:
> This is not gonna create annotations which do not coincide with the channel names as you wrote in the description, but it will completely ignore those annotations (and they will thus be absent from the Raw object entirely).

Actually I did not mention anywhere that these annotations would be created, but solely ignored 🤔 

> How come the recorded channel name is "EEG Fp1-CPz" and the annotation has a different channel name "Fp1"? Does that mean we erroneously parse the annotation read from an EDF file?

I suppose something like that happened:
1. The original .edf file was loaded with original montage where `"Fp1"` exists.
2. An annotation was added for channel `"Fp1"`.
3. The data was then re-referenced, transforming channel `"Fp1"` into `"EEG Fp1-CPz"`, _**but not modifying the annotation**_.
4. The data was then exported in a new .edf <= this is the file I suspect @tomasdominik is using

As per your initial in  comment https://github.com/mne-tools/mne-python/issues/12007#issuecomment-1734139304, for me the problem originates from the fact that either the re-referencing did not amend the annotations or that adding an annotation that did not match any channel did not trigger any error.

Reply:
In the above re-referencing example, what would you do with the annotation linked to `"Fp1"`? Would you:
1/ amend it to link it to `"EEG Fp1-CPz"`,
2/ delete it
3/ or leave it as it  is?

Reply:
@tomasdominik right, you mentioned that by default your EEG device records bipolar channels referenced to CP. Was your annotation manually added with eego then?

Reply:
| the original annotations won't be updated and will therefore potentially not coincide with new channels

I might have misread this sentence as "create annotations which do not coincide with the channel names", apologies.

With ANT Neuro, the caps are referenced to CPz, and CPz is absent from the file I think (so `EEG CPz-CPz` is absent)`? I do not use the EDF format, but I do remember seeing the naming schema EEG E1-Ref often.
IMO, the simplest is to detect annotations that are linked to a non-existing channel, emit a warning, and then set the annotation as global on all channels.

Reply:
I'm not using their softwares often, but I could try to export a couple of files to EDF to check the format of the EDF export spit by the eego software.

Reply:
> @tomasdominik right, you mentioned that by default your EEG device records bipolar channels referenced to CP. Was your annotation manually added with eego software then?

@paulroujansky No. There are two types of annotations - those related to the LPT triggers and those the system automatically generates related to when a channel exceeds the impedance threshold. I suspect this issue might be related to the latter.

Reply:
> IMO, the simplest is to detect annotations that are linked to a non-existing channel, emit a warning, and then set the annotation as global on all channels.

Agreed @mscheltienne -- @paulroujansky you okay with this, namely rather than discarding the channel-specific annotation with a `ch_name` that does not exist, just make it a global one? That way the user can in principle at least recover the channel-specific ones manually for example using `read_annotations` then doing some parsing at their own end.

Reply:
Actually I add a quick look into @tomasdominik's file and it seems that indeed

>the system automatically generates [annotations] related to when a channel exceeds the impedance threshold

![image](https://github.com/mne-tools/mne-python/assets/25418156/ba6857b6-656f-47d5-a0eb-5e27e897a2fd)

In this case, it's likely that the threshold was hit by all channels. Yet, in the case where it's not, we would definitely loose some information by removing the channel.
Perhaps what you proposed was more appropriate, namely propagate the annotation to new channels (here for instance `"EEG Fp1-CPz"`, then `"EEG Fpz-CPz"`, then `"EEG Fp2-CPz"` etc.), or leave as is.

Reply:
Channel-specific annotations are not specified in EDF. In fact, annotations are just arbitrary text. Therefore, I'd like to keep an option to load annotations without any processing on our side.

Also, it would be worth checking that file in EDFbrowser to see if it is standard compliant (very likely yes, but better make sure).

Reply:
> rather than discarding the channel-specific annotation with a ch_name that does not exist, just make it a global one

That seems appropriate, but I am also not using the EDF format often.

Reply:
> Channel-specific annotations are not specified in EDF. In fact, annotations are just arbitrary text. Therefore, I'd like to keep an option to load annotations without any processing on our side.

@cbrnr Actually it can be performed in EDF+, [as per the doc](https://www.edfplus.info/specs/edftexts.html), through a naming convention in the annotation label:

> **3.3. Linking annotations to signal channels**
Linking signals to specific recording channels can be optionally specified in EDF+ for which each corresponding annotation must be followed by the two-character string '@@', followed by the corresponding standard EDF+ label. Examples are: 'Limb movement@@EMG RAT' or 'EEG arousal@@Fpz-Cz' .

hence PR https://github.com/mne-tools/mne-python/pull/11960. Pretty handy when annotating a file at the channel-level (if channels exist though).

> Also, it would be worth checking that file in EDFbrowser to see if it is standard compliant (very likely yes, but better make sure).

For the info, the screenshot I passed hereabove that shows annotations comes from EDFbrowser after I loaded @tomasdominik's EDF file. We can see the @@ channel-convention applied here.

Reply:
IMO, it would perhaps be better to leave annotations as they are and raise a warning (as we do) so that the user can potentially amend annotations using his own logic.

Reply:
I agree with @paulroujansky, at least there should be a parameter to force that behavior.

Reply:
We could additionnaly implement a mechanism that describes what @larsoner proposed, namely convert channel-specific annotations to global ones in the case the channel does not exist, and an option to do that when loading a file (would probably need to drop duplicate global events afterwards). Do you think that would be worth it?

Reply:
This could be a rabbit hole, but I didn't know that channel-specific annotations were even defined in the specs (they are). I've never seen such files in the wild, but that doesn't mean anything. Maybe @Teuniz has an opinion on this issue?

But yeah, maybe converting to global annotations is the way to go. In that case, we could even append the offending `@@channel_label` to make it clear that this annotation is supposed to be channel-specific, but we couldn't parse it.

And in any case, if anything like this happens, we should issue a warning.

Reply:
> This could be a rabbit hole, but I didn't know that channel-specific annotations were even defined in the specs (they are). I've never seen such files in the wild, but that doesn't mean anything. Maybe @Teuniz has an opinion on this issue?

I know of at least one university & research center and a handful of other researchers who contacted me because they wanted to have some features in EDFbrowser related to signal linked annotations.  
  
E.g. "To quickly create a signal linked annotation by drawing a rectangle with the mouse, keep the Ctrl key pressed when releasing the left mouse button."  
"Alternatively, press 'a' when one or two crosshairs are enabled (on the same signal) in order to create a signal linked annotation."  
  
https://www.teuniz.net/edfbrowser/EDFbrowser%20manual.html#Annotation_editor  
  
I guess the reason that you don't usually see these signal linked annotations in EDF files in the wild,  
is because they use them internally in their organization.

Reply:
Thanks, yeah, that might be true. How would you deal with signal-specific annotations where the channel label does not exist in the data?